course6/lec6_script_en.md

```diff
@@ -0,0 +1,347 @@
+
+Programming languages provide us with various means of abstraction, such as functions, generics, higher-order functions, interfaces, etc. This chapter will introduce generics and higher-order functions, and the next chapter will discuss interfaces.
+
+# Generic Functions and Generic Data
+
+Let's first look at the stack data structure to understand why and how we use generics.
+
+A stack is a collection composed of a series of objects, where the insertion and removal of these objects follow the Last-In-First-Out (LIFO) principle. For example, consider the containers stacked on a ship as shown in the left-hand image below. 
+
+![](../pics/stack-objects.drawio.png)
+
+Clearly, new containers are stacked on top, and when removing containers, those on top are removed first, meaning the last placed container is the first to be removed. Similarly, with a pile of stones in the right-hand image, if you don’t want to topple the pile, you can only add stones at the top or remove the most recently added stones. This structure is a stack. There are many such examples in our daily lives, but we will not enumerate them all here.
+
+For a data type stack, we can define operations as follows. Taking an integer stack `IntStack` as an example, we can create a new empty stack; we can add an integer to the stack; we can try to remove an element from the stack, which may not exist because the stack could be empty, hence we use an Option to wrap it.
+
+```moonbit no-check
+empty: () -> IntStack // create a new stack
+push : (Int, IntStack) -> IntStack // add a new element to the top of the stack
+pop: IntStack -> (Option[Int], IntStack) // remove an element from the stack
+```
+
+As shown in the diagram below, we add a 2 and then remove a 2. We simply implement this definition of a stack.
+
+![](../pics/stack-push-pop-en.drawio.png)
+
+```moonbit
+enum IntStack {
+  Empty
+  NonEmpty(Int, IntStack)
+}
+fn IntStack::empty() -> IntStack { Empty }
+fn push(self: IntStack, value: Int) -> IntStack { NonEmpty(value, self) }
+fn pop(self: IntStack) -> (Option[Int], IntStack) {
+  match self {
+    Empty => (None, Empty)
+    NonEmpty(top, rest) => (Some(top), rest)
+  }
+}
+```
+
+In the code snippet, we see that we set the first argument as `IntStack`, and the variable name is `self`, allowing us to chain function calls. This means we can write `empty.push.pop` instead of `pop(push(empty))`. The deeper meaning of this syntax will be explained in the next chapter.
```

Comment:
```suggestion
In the code snippet, we see that we set the first argument as `IntStack`, and the variable name is `self`, allowing us to chain function calls. This means we can write `IntStack::empty().push(2).pop()` instead of `pop(push(2, IntStack::empty()))`. The deeper meaning of this syntax will be explained in the next chapter.
```